esp32/modules/services.py

```diff
@@ -1,59 +1,250 @@
     except OSError:
-        print("[SERVICES] Can't setup services: no lib folder!")
-        return False
-    found = False
+        return [False, False]
+    
+    #For each app...
     for app in apps:
+        print("APP: "+app)
         try:
-            files = uos.listdir('lib/'+app)
-        except OSError:
-            print("[SERVICES] Listing app files for app '"+app+"' failed!")
-
-        for f in files:
-            if (f=="service.py"):
-                found = True
-                print("[SERVICES] Found service "+app+"...")
-                try:
-                    srv = __import__('lib/'+app+'/service')
-                    services.append(srv) #Add to global list
-                    srv.setup()
-                except BaseException as msg:
-                    print("Exception in service setup "+app+": ", msg)
-                break
-    return found
+            #Try to open and read the json description
+            fd = open('/lib/'+app+'/service.json')
+            description = ujson.loads(fd.read())
+            fd.close()
+        except:
+            print("[SERVICES] No description found for "+app)
+            continue #Or skip the app
+        
+        print("DESC FOUND FOR: "+app)
+        
+        try:
+            #Try to open the service itself
+            fd = open('/lib/'+app+'/service.py')
+            fd.close()
+        except:
+            print("[SERVICES] No script found for "+app)
+            continue #Or skip the app
+        
+        print("SCRIPT FOUND FOR: "+app)
+        
+        rtcRequired = False # True if RTC should be set before starting service
+        loopEnabled = False # True if loop callback is requested
+        drawEnabled = False # True if draw callback is requested
+        
+        wifiInSetup = False # True if wifi needed in setup
+        wifiInLoop = False # True if wifi needed in loop
+        
+        try:
+            if description['apiVersion']!=1:
+                print("[SERVICES] Service for "+app+" is not compatible with current firmware")
+                continue #Skip the app
+            wifiInSetup = description['wifi']['setup']
+            wifiInLoop = description['wifi']['setup']
+            rtcRequired = description['rtc']
+            loopEnabled = description['loop']
+            drawEnabled = description['draw']
+        except:
+            print("[SERVICES] Could not parse description of app "+app)
+            continue #Skip the app
+        
+        print("[SERVICES] Found service for "+app)
+        
+        # Import the service.py script
+        try:
+            srv = __import__('lib/'+app+'/service')
+        except:
+            print("[SERVICES] Could not import service of app "+app)
+            continue #Skip the app
+        
+        if wifiInSetup:
+            if wifiFailed:
+                print("[SERVICES] Service of app "+app+" requires wifi and wifi failed so the service has been disabled.")
+                continue
```

Comment:
I agree that this should be improved, in the future.

Workaround: For now just set the wifi flag to false in the service.json file and manually call easywifi.connect(). Apps can use the return value to determine what to do. This does result in multiple apps trying to connect to wifi on their own accord, draining battery.